File: perl/parser.pl
#!/usr/bin/perl

my $filename = $ARGV[0];

# read translation text file
my $file = $filename; #"../translations/NIV.txt"; #"$ARGV[1];
my $document = do {
    local $/ = undef;
    open my $fh, "<", $filename
        or die "could not open $filename: $!";
    <$fh>;
};

$numSentences = 0;
$numWords = 0;
$numSyllables = 0;

# count occurances of each sentence end punctuation
my $periods = () = ($document =~ /\./g);
my $colons = () = ($document =~ /:/g);
my $semicolons = () = ($document =~ /;/g);
my $questions = () = ($document =~ /\?/g);
my $exclamations = () = ($document =~ /!/g);

$numSentences = $periods + $colons + $semicolons + $questions + $exclamations;


# split string with entire translation into individual word array
my @wordArray = split ' ', $document;

$prevVowel = 0;

foreach my $n (@wordArray){

    if(! ($n =~ /^\d+?$/)){

        # increment word count
        $numWords = 1 + $numWords;

        # remove punctuation from word
        $n =~ s/[[:punct:]]//g;

        # set word to lower case
        $n = lc $n;

        #print "$n\n";

        my @chars = split("", $n);
        $wordSyllables = 0;
        $prevVowel = 0;

        #count syllables in each word
        foreach my $c (@chars){
            if($c eq 'a' || $c eq 'e' || $c eq 'i' || $c eq 'o' || $c eq 'u' || $c eq 'y'){
                if($prevVowel == 0){
                    $wordSyllables = $wordSyllables + 1;              
                }  
                $prevVowel = 1;
            } 
            else{
                $prevVowel = 0;
            }
        }

        #handle silent e
        if($chars[@chars-1] eq 'e'){
            if($wordSyllables > 1){
                $wordSyllables = $wordSyllables - 1;
            }
        }

        $numSyllables = $numSyllables + $wordSyllables;
    }
}

#print "$numWords\n";
#print "$numSyllables\n";
#print "$numSentences\n";

# calculate and output Flesch Indeces

$alpha = $numSyllables/$numWords;
$beta = $numWords/$numSentences; 

my $fleschIndex = 206.835 - $alpha*84.6 - $beta*1.015;
my $fleschKincaidIndex = $alpha*11.8 + $beta*0.39 - 15.59;

print "Flesch Readability Index: $fleschIndex\n";
print "Flesch-Kincaid Grade Level Index: $fleschKincaidIndex\n";

File: fortran/parser.f90
program reader2

character(LEN=5000000)::long_string
integer::filesize

! read file into one long_string
call read_file(long_string, filesize)

end program reader2

subroutine read_file(string, counter)
character(LEN=*)::string
character(LEN=32)::arg
integer::counter
character(LEN=1)::input, prev
real::wordCount
real::sentenceCount
real::syllablesCount, wordSyllables
logical:: sentenceMarker, vowelIn, vowelPrev

wordCount = 0.0
sentenceCount = 0.0
syllablesCount = 0.0
prev = " "

call getarg(1, arg)

open (unit=5, status = "old",access="direct",form="unformatted",recl=1,file=trim(arg))

counter=1
wordSyllables = 0.0
100 read (5, rec=counter,err=200) input
    counter=counter+1
    call isSentenceMarker(input, sentenceMarker)
    call clean_char(input)
    
    !increment sentence count
    if(sentenceMarker) then
        sentenceCount = sentenceCount + 1
    end if

    !increment word count based on spaces following chars 
    if(input /= " " .and. prev == " ") then
        wordCount = wordCount + 1.0
    end if

    call isVowel(input, vowelIn)

    ! count syllables
    if(input == " ") then
        ! handle silent e
        if(prev == "E") then 
            if(wordSyllables == 0.0)then
                wordSyllables = 1.0;
            else if(wordSyllables > 1.0) then
                wordSyllables = wordSyllables - 1.0;
            end if
        end if
        syllablesCount = wordSyllables + syllablesCount
        wordSyllables = 0
    else
        ! base case 
        if(vowelIn .and. .not. vowelPrev) then
            wordSyllables = wordSyllables + 1
        end if
    end if

    vowelPrev = vowelIn
    prev = input
    string(counter:counter)=input
    goto 100
200 continue
counter=counter-1

close(5)

! calculate and print Flesch Indeces
call getFleschIndex(wordCount, sentenceCount, syllablesCount)
call getFleschKincaidIndex(wordCount, sentenceCount, syllablesCount)

end subroutine read_file


subroutine getFleschIndex(wordCount, sentenceCount, syllablesCount)
! calculates and prints Flesch Index
real::wordCount, sentenceCount, syllablesCount, alpha, beta, fIndex
alpha = syllablesCount/wordCount
beta = wordCount/sentenceCount    
fIndex = 206.835 - alpha*84.6 - beta*1.015
print*, "Flesch Readability Index: ", fIndex

end subroutine getFleschIndex


subroutine getFleschKincaidIndex(wordCount, sentenceCount, syllablesCount)
! calculates and prints Flesch Kincaid Index
real::wordCount, sentenceCount, syllablesCount, alpha, beta, fIndex
alpha = syllablesCount/wordCount
beta = wordCount/sentenceCount
fIndex = alpha*11.8 + beta*0.39 - 15.59
print*, "Flesch-Kincaid Grade Level Index: ", fIndex
end subroutine getFleschKincaidIndex


subroutine isVowel(charNext, vowel)
! checks if word is vowel
character(len=1) :: charNext
logical :: vowel 
vowel = (charNext == "A" .or. charNext == "E" .or. charNext == "I" .or. charNext == "O" .or. charNext == "U" .or. charNext == "Y")
end subroutine isVowel

subroutine isSentenceMarker(charNext, sentenceMarker)
! check for presence of sentence markers
character(len=1) :: charNext
logical :: sentenceMarker
sentenceMarker = (charNext == "." .or. charNext == ":" .or. charNext == ";" .or. charNext == "!" .or. charNext == "?")
end subroutine isSentenceMarker

subroutine clean_char(charNext)
! removes all non-characters and sets to uppercase
character(len=1) :: charNext
integer :: j, e
logical :: is_numeric
real :: x
j = iachar(charNext(1:1))
if(j>=iachar("a") .and. j<=iachar("z")) then 
    charNext(1:1) = achar(iachar(charNext(1:1))-32)
end if

if((j<iachar("A"))) then
  charNext(1:1) = " "
end if

if(j > ichar("Z") .and. j < 61) then
    charNext(1:1) = " "
end if

if(charNext == "]" .or. charNext == "[") then
    charNext(1:1) = " "
end if

read(charNext,*, iostat = e) x
is_numeric = e == 0
if(is_numeric ) then
    charNext(1:1) = " "
end if
end subroutine clean_char
